app/scripts/modules/core/serverGroup/serverGroupWriter.service.ts

```diff
@@ -0,0 +1,207 @@
+import {module} from 'angular';
+
+import {IJob} from 'core/application/service/application.write.service';
+import {ITask} from 'core/task/task.read.service';
+import {ServerGroup, ISecurityGroup} from 'core/domain';
+import {TASK_EXECUTOR, TaskExecutor} from 'core/task/taskExecutor';
+
+interface IServerGroupCommandApplication {
+  name: string;
+}
+
+interface ISource {
+  asgName: string;
+}
+
+interface IViewState {
+  mode: string;
+}
+
+interface IServerGroupTaskApplicationTask {
+  refresh?: () => void;
+}
+
+interface ICapacity {
+  desired: number;
+  max: number;
+  min: number;
+}
+
+export interface IServerGroupCommand {
+  application: IServerGroupCommandApplication;
+  freeFormDetails?: string;
+  source?: ISource;
+  stack?: string;
+  type?: string;
+  viewState: IViewState;
+}
+
+export interface IServerGroupTaskApplication {
+  name: string;
+  tasks?: IServerGroupTaskApplicationTask;
+}
+
+export interface IServerGroupJob extends IJob {
+  amiName?: string;
+  asgName?: string;
+  capacity?: ICapacity;
+  credentials?: string;
+  cloudProvider?: string;
+  region?: string;
+  securityGroups?: string[];
+  serverGroupName?: string;
+}
+
+export class ServerGroupWriterService {
+
+  static get $inject(): string[] {
+    return ['taskExecutor', 'serverGroupTransformer'];
+  }
+
+  constructor(private executor: TaskExecutor, private transformer: any) {}
+
+  public cloneServerGroup(command: IServerGroupCommand,
+                          application: IServerGroupTaskApplication): ng.IPromise<ITask> {
+
+    let description: string;
+    if (command.viewState.mode === 'clone') {
+      description = `Create Cloned Server Group from ${command.source.asgName}`;
+      command.type = 'cloneServerGroup';
+    } else {
+      command.type = 'createServerGroup';
+      let asgName = application.name;
+      if (command.stack) {
+        asgName += `-${command.stack}`;
+      }
+      if (!command.stack && command.freeFormDetails) {
+        asgName += '-';
+      }
+      if (command.freeFormDetails) {
+        asgName += `-${command.freeFormDetails}`;
+      }
```

Comment:
Since we're refactoring this service, it's probably better to call `namingService.getClusterName(application.name, command.stack, command.freeFormDetails)` to get the cluster name (and rename the variable `clusterName`, since it's not really the name of the server group).